src/lib.rs

```diff
@@ -38,6 +38,79 @@ pub mod tpm;
 // future goal ... once I can afford one ...
 // mod yubihsm;
 
+pub(crate) const TPM_PIN_MIN_LEN: u8 = 6;
+// TPM's limit the max pin based on algorithm max bytes per the
+// size of the largest hash. This means pins max out at 32 bytes
+// as that's the size of sha256 output.
+pub(crate) const TPM_PIN_MAX_LEN: u8 = 32;
+
+pub struct PinValue {
+    value: Zeroizing<Vec<u8>>,
+}
+
+#[derive(Debug)]
+pub enum TpmPinError {
+    TooShort(u8),
+    TooLarge(u8),
+}
+
+impl PinValue {
+    pub fn new(input: &str) -> Result<Self, TpmPinError> {
+        if input.len() < TPM_PIN_MIN_LEN as usize {
+            return Err(TpmPinError::TooShort(TPM_PIN_MIN_LEN));
+        } else if input.len() > TPM_PIN_MAX_LEN as usize {
+            return Err(TpmPinError::TooLarge(TPM_PIN_MAX_LEN));
+        }
+
+        Ok(PinValue {
+            value: input.as_bytes().to_vec().into(),
+        })
+    }
+
+    pub(crate) fn derive_aes_256_gcm(
+        &self,
+        parent_key: &[u8],
+    ) -> Result<Zeroizing<[u8; AES256GCM_KEY_LEN]>, TpmError> {
+        use argon2::{Algorithm, Argon2, Params, Version};
+
+        let mut auth_key = Zeroizing::new([0; AES256GCM_KEY_LEN]);
+
+        // This can't be changed else it will break key derivation for users.
+        let argon2id_params =
+            Params::new(32_768, 1, 1, Some(auth_key.as_ref().len())).map_err(|argon_err| {
+                error!(?argon_err);
+                TpmError::AuthValueDerivation
+            })?;
+
+        // Want at least 8 bytes salt, 16 bytes pw input.
+        if parent_key.len() < 24 {
+            return Err(TpmError::AuthValueTooShort);
+        }
+
+        let (salt, pepper) = parent_key.split_at(MIN_SALT_LEN);
+
+        let argon =
+            Argon2::new_with_secret(pepper, Algorithm::Argon2id, Version::V0x13, argon2id_params)
+                .map_err(|argon_err| {
+                error!(?argon_err);
+                TpmError::AuthValueDerivation
+            })?;
+
+        let now = std::time::SystemTime::now();
+
+        argon
+            .hash_password_into(self.value.as_ref(), salt, auth_key.as_mut())
+            .map_err(|argon_err| {
+                error!(?argon_err);
+                TpmError::AuthValueDerivation
+            })?;
+
+        error!(elapsed = ?now.elapsed());
```

Comment:
Looks like it's coming from here. Perhaps this is just a testing statement you forgot to remove?

Reply:
Yeah, it is 🤣